File: Task4/src/ForAll.scala
/**
 * Created by Шолохов on 30.01.2015.
 */
class ForAll(override val variable : Variable, override  val expression : Expression) extends Quantifier (variable, expression){
  override def toString(): String = "@" + super.toString()
  override def equals(o : Any): Boolean = o.isInstanceOf[ForAll] && super.equals(o)
}

File: Task4/src/Forall.java
import java.util.HashMap;

public class Forall extends Statement {
    public final Statement child;
    public final String varName;

    public Forall(String varName, Statement child) {
        this.varName = varName;
        this.child = child;
    }

    protected boolean compareImpl(String[] patterns, Statement other) {
        if (other instanceof Forall && varName.equals(((Forall)other).varName)) {
            return child.compareWithinContext(patterns, ((Forall)other).child);
        } else {
            return false;
        }
    }

    public Statement substituteTerm(Expression where, Expression what) {
        return new Forall(varName, child.substituteTerm(where, what));
    }

    public Statement substitutePatterns(Statement[] to) {
        return new Forall(varName, child.substitutePatterns(to));
    }

    public boolean estimate(HashMap<String, Boolean> values) {
        return false;
    }

    public String toString() {
        return "@" + varName + child;
    }
}
